workflow/controller/controller.go

```diff
@@ -553,37 +554,66 @@ func (wfc *WorkflowController) workflowGarbageCollector(stopCh <-chan struct{})
 					log.WithField("err", err).Error("Failed to list old offloaded nodes")
 					continue
 				}
-				if len(oldRecords) == 0 {
-					log.Info("Zero old offloads, nothing to do")
-					continue
-				}
-				// get every lives workflow (1000s) into a map
-				liveOffloadNodeStatusVersions := make(map[types.UID]string)
-				workflows, err := util.NewWorkflowLister(wfc.wfInformer).List()
-				if err != nil {
-					log.WithField("err", err).Error("Failed to list incomplete workflows")
-					continue
+				lenOldOffloads := 0
+				for _, records := range oldRecords {
+					lenOldOffloads += len(records)
 				}
-				for _, wf := range workflows {
-					// this could be the empty string - as it is no longer offloaded
-					liveOffloadNodeStatusVersions[wf.UID] = wf.Status.OffloadNodeStatusVersion
-				}
-				log.WithFields(log.Fields{"len_wfs": len(liveOffloadNodeStatusVersions), "len_old_offloads": len(oldRecords)}).Info("Deleting old offloads that are not live")
-				for _, record := range oldRecords {
-					// this could be empty string
-					nodeStatusVersion, ok := liveOffloadNodeStatusVersions[types.UID(record.UID)]
-					if !ok || nodeStatusVersion != record.Version {
-						err := wfc.offloadNodeStatusRepo.Delete(record.UID, record.Version)
-						if err != nil {
-							log.WithField("err", err).Error("Failed to delete offloaded nodes")
-						}
+				log.WithFields(log.Fields{"len_wfs": len(oldRecords), "len_old_offloads": lenOldOffloads}).Info("Deleting old offloads that are not live")
+				for uid, versions := range oldRecords {
+					if err := wfc.deleteOffloadedNodesForWorkflow(uid, versions); err != nil {
+						log.WithError(err).WithField("uid", err).Error("Failed to delete old offloaded nodes")
 					}
 				}
+				log.Info("Workflow GC finished")
 			}
 		}
 	}
 }
 
+func (wfc *WorkflowController) deleteOffloadedNodesForWorkflow(uid string, versions []string) error {
+	workflows, err := wfc.wfInformer.GetIndexer().ByIndex(indexes.UIDIndex, uid)
+	if err != nil {
+		return err
+	}
+	var wf *wfv1.Workflow
+	switch l := len(workflows); l {
+	case 0:
+		log.WithField("uid", uid).Info("Workflow missing, probably deleted")
+	case 1:
+		un := workflows[0].(*unstructured.Unstructured)
+		wf, err = util.FromUnstructured(un)
+		if err != nil {
+			return err
+		}
+		key, err := cache.MetaNamespaceKeyFunc(wf)
```

Comment:
minor - just do `namespace+"/"+name` - simpler and no error check needed

Reply:
Are you sure? Everywhere else in the controller cache.MetaNamespaceKeyFunc is being used to compute the key. Isn't duplicating the behavior of this function a maintenance hazard? What if cache.MetaNamespaceKeyFunc is changed in the future?

Reply:
Good point, but I don't expect that to happen in reality. Too much code depends on it.

Reply:
See `cache.ExplicitKey`